cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemotePrivilegesHandler.java

```diff
@@ -49,15 +46,25 @@
   private static final Logger LOG = LoggerFactory.getLogger(RemotePrivilegesHandler.class);
   private static final Type SET_OF_ACTIONS = new TypeLiteral<Set<Action>>() { }.getType();
 
-  private final Authorizer authorizer;
-  private final PrivilegesFetcher privilegesFetcher;
+  private final PrivilegesManager privilegesManager;
+  private final AuthorizationEnforcer authorizationEnforcer;
 
   @Inject
-  RemotePrivilegesHandler(
-    AuthorizerInstantiator authorizerInstantiator,
-    @Named(AuthorizationEnforcementModule.PRIVILEGES_FETCHER_PROXY_CACHE) PrivilegesFetcher privilegesFetcher) {
-    this.authorizer = authorizerInstantiator.get();
-    this.privilegesFetcher = privilegesFetcher;
+  RemotePrivilegesHandler(PrivilegesManager privilegesManager, AuthorizationEnforcer authorizationEnforcer) {
+    this.privilegesManager = privilegesManager;
+    this.authorizationEnforcer = authorizationEnforcer;
+  }
+
+  @POST
+  @Path("/enforce")
+  public void enforce(HttpRequest request, HttpResponder responder) throws Exception {
+    Iterator<MethodArgument> arguments = parseArguments(request);
+    EntityId entityId = deserializeNext(arguments);
+    Principal principal = deserializeNext(arguments);
```

Comment:
The key for the cache is Principal, Entity, and Action, so we need to enforce on individual actions. The reason behind this is that in the case of an operation requiring multiple privileges we should be able to tell the user which privilege they are missing. This method is called by the cache's load method in the case of a cache miss and that only has access to the key class, so this enforce will not be called with a Set. The reason to have two classes is that we don't want to force the user to grant and revoke on single actions and continue supporting the multi-action commands but we need to call enforce Action be Action.